### DISCUSSION TITLE
Adding standard iso8601 json decoder

### DISCUSSION THREAD

**Discussion by JaapWijnen** [pointfreeco/swift-url-routing#24] [Ideas]
Some discussion is probably needed on how to best approach this. Today I ran into the issues of one of my routes not working when using swift-url-routing in combination with Vapor's testing capabilities. After some digging I found it was due to Vapor's JSONDecoder and encoder defaulting to a .iso8601 date decoding/encoding strategy. 
This means in our routers whenever we're using `Date` objects we'd have to inject custom json encoders and furthermore for first time users it's quite difficult to pinpoint where the issue originates from. 
Ideally we could do some kind of addition that lives in `vapor-routing` to not pollute this package too much with Vapor implementation details. But I'm not sure of what a good solution would be.

**Comment by stephencelis** [comment]
@JaapWijnen We have an idea of how to address this problem, but it unfortunately depends on some Swift 5.7 features to do it right.

**Reply by JaapWijnen** [reply]
Cool! I'd be interested to hear about it

**Reply by stephencelis** [reply]
The short of it is we need a way of propagating dependencies through the router tree. We can do so in a fashion similar to SwiftUI's `@Environment`, but it requires [primary associated types](https://github.com/apple/swift-evolution/blob/main/proposals/0346-light-weight-same-type-syntax.md) to be ergonomic. Something like:

```swift
struct MyRouter: ParserPrinter {
  var body: some ParserPrinter<URLRequestData, MyRoute> {
    ParsePrint {
      ...
    }
    // propagate custom JSON coding to all parsers defined above
    .json(decoder: customDecoder, encoder: customEncoder)
  }
}
```

**Comment by stephencelis** [comment]
I'm going to convert this to a discussion for now, since it's not a "bug" with the library per se, and because we have plans for how to address more holistically in the future. For now as a workaround we recommend manually passing custom decoders/encoders.
